packages/lesswrong/server/migrations/2020-05-06-forceSecureImageLinks.ts

```diff
@@ -0,0 +1,112 @@
+/*
+ * Turn insecure image links into secure ones if the secure image exists. We
+ * only edit the html (not the original content), so it's possible someone will
+ * edit their post and the update will be reverted.
+ */
+
+import { registerMigration, forEachDocumentBatchInCollection } from './migrationUtils';
+import { urlIsBroken } from '../scripts/utils'
+import { Posts } from '../../lib/collections/posts/collection';
+import cheerio from 'cheerio'
+
+function findInsecureImages ($: any): Array<string> {
+  let insecureImageSources: Array<string> = []
+  $('img').each(function (this: any) {
+    const el = $(this)
+    const src = el.attr('src') as string
+    const testResult = /^http:\/\//.test(src)
+    if (testResult) {
+      insecureImageSources.push(src)
+    }
+  })
+  return insecureImageSources
+}
+
+async function testSecureImages
+  (insecureImageSources: Array<string>, post: DbPost):
+  Promise<{shouldUpdate: boolean, imageUpdates: Map<string, string>}>
+{
+  let shouldUpdate = false
+  const imageUpdates = new Map((await Promise.all(
+    insecureImageSources.map(async (src): Promise<[string, string | boolean]> => {
+      const secureImageSource = src.replace(/^http:/, 'https:')
+      if (!(await urlIsBroken(secureImageSource))) {
+        shouldUpdate = true
+        return [src, secureImageSource]
+      }
+      return [src, false]
+    })
+  )).filter(tuple => tuple[1])) as Map<string, string>
+  return {shouldUpdate, imageUpdates}
+}
+
+function updateHtmlWithSecureImages($: any, imageUpdates: Map<string, string>): void {
+  $('img').each(function (this: any) {
+    const el = $(this)
+    const src = el.attr('src') as string
+    const update = imageUpdates.get(src)
```

Comment:
I'm not sure I love this two-pass pattern, where you first extract all the links to be fixed, and then re-scan the HTML and update the source. I assume there's a good reason for it, but it seems a bit janky and it's hard to reason through.

Why not just write it as 

```js

 $('img').each(function (this: any) {
    const el = $(this)
    const src = el.attr('src') as string
	const secureSrc = getSecureSrc(src)
	if (src !== secureSrc) {
		el.attr('src', secureSrc)
		return
	}
  })
```

and then have a `getSecureSrc()` function that does your regex and broken link tests?

Thinking about it more, maybe it's that it's hard to run the async operations within the Cheerio callback?

Reply:
Ah, I guess the two-pass pattern also gives you the ability to test `shouldUpdate` more easily in `getFixedHTML()`...

Reply:
Nope, it's because the cheerio each api doesn't let me do async functions. So I needed to have an async part in between the find and update cheerio steps.